genesis/utils/usd.py

```diff
@@ -0,0 +1,312 @@
+
+        diffuse_texture = diffuse_pattern.group(1) if diffuse_pattern else None
+        normal_texture = normal_pattern.group(1) if normal_pattern else None
+        orm_texture = orm_pattern.group(1) if orm_pattern else None
+
+        if diffuse_tint_pattern:
+            # Extract the comma-separated color values from inside the parentheses
+            tint_values_str = diffuse_tint_pattern.group(1)
+            # Convert each value to a float and form a tuple
+            diffuse_tint = tuple(float(x.strip()) for x in tint_values_str.split(","))
+        else:
+            diffuse_tint = None
+
+        return diffuse_texture, normal_texture, diffuse_tint, orm_texture
+
+    """parse mesh from USD file"""
+    meshes = list()
+    stage = Usd.Stage.Open(path)
+    default_prim_path = stage.GetDefaultPrim().GetPath().pathString
+    for prim in stage.Traverse():
+        # only load mesh under the default prim
+        if UsdGeom.Mesh(prim) and (default_prim_path in prim.GetPath().pathString):
+            # check surface
+            mesh_surface = gs.surfaces.Default() if surface is None else surface.copy()
+
+            # load mesh basic information
+            mesh = UsdGeom.Mesh(prim)
+            points = mesh.GetPointsAttr().Get()
+            normals = mesh.GetNormalsAttr().Get()
+            face_vertex_counts = mesh.GetFaceVertexCountsAttr().Get()
+            face_vertex_indices = mesh.GetFaceVertexIndicesAttr().Get()
+            sts = mesh.GetPrim().GetAttribute("primvars:st").Get()
+            st_indices = mesh.GetPrim().GetAttribute("primvars:st:indices").Get()
+
+            # calculate transform
+            transform = np.array(get_world_transform(mesh.GetPrim()))
+            points_homogeneous = np.hstack([points, np.ones((len(points), 1))])
+
+            verts = (points_homogeneous @ transform)[:, :3]
+            normals = np.array(normals, dtype=float)
+            indices = np.array(face_vertex_indices, dtype=int)
+
+            faces = []
+            index_offset = 0
+            face_count = {}
+
+            for count in face_vertex_counts:
+                face_count[count] = face_count.get(count, 0) + 1
+                face = indices[index_offset : index_offset + count].tolist()
+
+                # If the face is a triangle, keep it as is
+                if count == 3:
+                    faces.append(face)
+                # If the face is a quad or more, triangulate it
+                elif count > 3:
+                    for i in range(1, count - 1):
+                        faces.append([face[0], face[i], face[i + 1]])  # Fan triangulation
+
+                index_offset += count  # Move to the next set of indices
+
+            # if face count is not 3, means it is not all triangle faces
+            if len(face_count) >= 2 or (3 not in face_count):
+                gs.logger.warning(f"Mesh {mesh} has non-triangle faces. Face count {face_count}")
```

Comment:
This this warning if you are doing triangulation anyway ? Just debug explaining the process should be fine.